### DISCUSSION TITLE
[React] Add React 19 support

### DISCUSSION THREAD

**Discussion by gperdomor** [nrwl/nx#29243] [Feature Requests]
### Current Behavior

React generators use React 18.3.0

### Expected Behavior

React generators should use React 19.0.0

### Steps to Reproduce

N/A

### Nx Report

N/A

### Operating System

- [ ] macOS
- [ ] Linux
- [ ] Windows
- [ ] Other (Please specify)

**Comment by jaysoo** [comment]
We are going to include it in the next 20.3 version, so new workspaces will use React 19 and Next.js 15. We will not auto-migrate workspaces that are using older versions since that will break apps.

You are free to upgrade to React 19 now. If you run into any Nx specific issues please report them.

I'm converting this to a discussion as it is not a bug.

**Reply by yoanne2x** [reply]
When do we expect the migration to be done?

**Reply by alterhuman** [reply]
20.3.3 is already here but no react 19

**Comment by SoftwareDesign-Solution** [comment]
I found a workaround. Before you create a React application with `npx nx g @nx/react:app apps/client`, you have to run the two commands first, then the generator will not install version 18.

```bash
npm install react react-dom
npm install --save-dev @types/react @types/react-dom
```

**Reply by pratik149** [reply]
Any idea how to do it in an existing react nx project? I'm planning to update from 18 to 19.

**Reply by rostyslav-nabok** [reply]
Just follow the React migration guide, migration from 18 to 19 does not relate to NX

**Reply by travisfranklin** [reply]
If you wish to continue using `@nx/react` as a generator while using react 19 you end up with peer deps issues. This is due to @module-federation/node not currently supporting React v19.

```
 WARN  Issues with peer dependencies found

├─┬ @nx/react 20.4.6
│ └─┬ @nx/module-federation 20.4.6
│   └─┬ @module-federation/node 2.6.28
│     ├── ✕ unmet peer react@^16||^17||^18: found 19.0.0
│     ├── ✕ unmet peer react-dom@^16||^17||^18: found 19.0.0
│     └─┬ @module-federation/utilities 3.1.46
│       ├── ✕ unmet peer react@"^16 || ^17 || ^18": found 19.0.0
│       └── ✕ unmet peer react-dom@"^16 || ^17 || ^18": found 19.0.0
```

In testing I haven't run into issues beyond the annoying warning, and having to generate and then update to 19, but most of my work is in nextjs currently.

**Comment by Aakarshak1** [comment]
Any update on React 19 support?

**Reply by Lonli-Lokli** [reply]
It is already supported

**Reply by Aakarshak1** [reply]
@Lonli-Lokli I just created a new workspace with the below configuration and it installed react v18.3.1 by default 
![Screenshot 2025-03-12 at 23 03 14](https://github.com/user-attachments/assets/4e37facf-cd1e-4bbb-b68d-7cd11d1c42a1)

Do I need to change anything here?

**Reply by SoftwareDesign-Solution** [reply]
> @Lonli-Lokli I just created a new workspace with the below configuration and it installed react v18.3.1 by default ![Screenshot 2025-03-12 at 23 03 14](https://private-user-images.githubusercontent.com/16081512/422010747-4e37facf-cd1e-4bbb-b68d-7cd11d1c42a1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE4MDE0NjUsIm5iZiI6MTc0MTgwMTE2NSwicGF0aCI6Ii8xNjA4MTUxMi80MjIwMTA3NDctNGUzN2ZhY2YtY2QxZS00YmJiLWI2OGQtN2NkMTFkMWM0MmExLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzEyVDE3MzkyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAyMTdiNmMzMjA2ZDc2Njc5MGI1M2Q1OWFjYzIxNDg0ODFhYjBmNGNkNDJhNmQ1NmJjODk1NjY5NWY5YjM4MWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.S58i9rNZNsK_FLJ4sEdDS7tPOsa2i1v6ItIymTehk0k)
> 
> Do I need to change anything here?

I have the same problem and used the following workaround.

Create an empty Nx monorepo and then run the following commands before adding @nx/react to the project.

`npm install react react-dom`

`npm install --save-dev @types/react @types/react-dom`

Die Überprüfung auf React 18 oder eine andere Version habe ich an folgende Stellen gefunden

https://github.com/nrwl/nx/blob/master/packages/react/src/utils/version-utils.ts

**Comment by cheesestringer** [comment]
For anyone blocked by the `@module-federation/node` package this was fixed in https://github.com/module-federation/core/pull/3619 and is now available in release https://github.com/module-federation/core/releases/tag/v0.11.2.

Do an `npm update` to consume the new versions.

**Reply by Juuldamen** [reply]
FYI, if you have the below error. While `@module-federation/node` `peerDependencies` have been updated, `@module-federation/utilities` has not been. It is waiting on https://github.com/module-federation/core/pull/3630 to resolve this error.

```bash
npm warn Could not resolve dependency:
npm warn peerOptional react@"^16 || ^17 || ^18" from @module-federation/utilities@3.1.49
```

**Reply by cheesestringer** [reply]
Utilities was just a warning so it shouldn't block you but maybe my setup was simpler. Otherwise yeah wait for the next release.
